### PR TITLE
Persistence Component: Interface for Multi Dataset Ingestion

### PR DESCRIPTION
**Pull request by kumuwu**
#### What type of PR is this?

- Improvement

#### What does this PR do / why is it needed ?

Provide new interface for multi-dataset ingestion

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/ApiUtils.java [hunk 1]
```diff
@@ -21,53 +21,89 @@
...
 
-    public static Dataset deriveMainDatasetFromStaging(Datasets datasets, IngestMode ingestMode)
+    public static Dataset deriveMainDatasetFromStaging(Dataset mainDataset, Dataset stagingDataset, IngestMode ingestMode)
     {
-        Dataset mainDataset = datasets.mainDataset();
-        List<Field> mainDatasetFields = mainDataset.schema().fields();
-        if (mainDatasetFields == null || mainDatasetFields.isEmpty())
+        Dataset enrichedMainDataset = mainDataset;
+        if (mainDataset instanceof DatasetReference ||
+                (mainDataset instanceof DatasetDefinition && mainDataset.schema().fields() == null || mainDataset.schema().fields().isEmpty()))
```
Comment (prasar-ashutosh) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/ApiUtils.java:102]: Or conditions should be in brackets:
false && true || true => true

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/ApiUtils.java [hunk 2]
```diff
@@ -184,6 +220,221 @@ private static LockInfoDataset getLockInfoDataset(Datasets datasets)
...
+        {
+            List<Map<String, Object>> metadataResults = tabularDataList.get(0).getData();
+            for (Map<String, Object> metadata: metadataResults)
+            {
+                Timestamp ingestionTimestampUTC = (Timestamp) metadata.get(metadataDataset.batchStartTimeField());
+                Timestamp ingestionEndTimestampUTC = (Timestamp) metadata.get(metadataDataset.batchEndTimeField());
+                String batchStatus = String.valueOf(metadata.get(metadataDataset.batchStatusField()));
+                batchStatus = batchStatus.equalsIgnoreCase(batchSuccessStatusValue)  ? IngestStatus.SUCCEEDED.name() : batchStatus;
+                IngestorResult ingestorResult = IngestorResult.builder()
+                    .batchId((int) metadata.get(metadataDataset.tableBatchIdField()))
```
Comment (prasar-ashutosh) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/ApiUtils.java:247]: Casting to int is creating problems as some times it's coming as long

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/util/BatchErrorDatasetUtils.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    private final BatchErrorDataset batchErrorDataset;
+    private final Dataset dataset;
+
+    public BatchErrorDatasetUtils(BatchErrorDataset batchErrorDataset, Dataset dataset)
+    {
+        this.batchErrorDataset = batchErrorDataset;
+        this.dataset = dataset;
+    }
+
+    public Insert insertBatchError(String requestId, String tableName, String errorMessage, String errorCategory, BatchStartTimestamp batchStartTimestamp)
```
Comment (prasar-ashutosh) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/util/BatchErrorDatasetUtils.java:37]: For Ashutosh -- Bring Error Category and Error Sub Category in Lib

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalMultiDatasetIngestorAbstract.java [hunk 4]
```diff
@@ -0,0 +1,642 @@
...
+        }
+        else
+        {
+            return stagingDataset;
+        }
+    }
+
+    private Dataset deriveMainDataset(IngestStage ingestStage, IngestMode enrichedIngestMode)
+    {
+        DatasetReference mainDatasetReference = ingestStage.mainDataset();
```
[the remark this answers is not in the document]
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalMultiDatasetIngestorAbstract.java:612]: As discussed, calling directly will cause runtime errors, so leaving as it is

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/ApiUtils.java (hunk 1 above)
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/ApiUtils.java]: No longer applicable

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalMultiDatasetIngestorAbstract.java [hunk 5]
```diff
@@ -0,0 +1,642 @@
...
+                    .stagingDataset(deriveStagingDataset(ingestStage))
+                    .mainDataset(deriveMainDataset(ingestStage, ingestStage.ingestMode()))
+                    .metadataDataset(metadataDataset)
+                    .build();
+
+                // 2. Enrich the ingest mode with case conversion
+                IngestMode enrichedIngestMode = ApiUtils.applyCase(ingestStage.ingestMode(), caseConversion());
+
+                // 3. Enrich the datasets with case conversion
+                enrichedDatasets = ApiUtils.enrichAndApplyCase(enrichedDatasets, caseConversion());
```
[the remark this answers is not in the document]
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalMultiDatasetIngestorAbstract.java]: Sure, changed!

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/ingestmode/NoOpAbstract.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
+// Copyright 2022 Goldman Sachs
```
Comment (kumuwu) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/ingestmode/NoOpAbstract.java:1]: typo

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/NoOpPlanner.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
+// Copyright 2022 Goldman Sachs
```
Comment (kumuwu) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/NoOpPlanner.java:1]: typo

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-ansi/src/test/java/org/finos/legend/engine/persistence/components/ingestmode/NoOpTest.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
+// Copyright 2022 Goldman Sachs
```
Comment (kumuwu) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-ansi/src/test/java/org/finos/legend/engine/persistence/components/ingestmode/NoOpTest.java:1]: typo

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-h2/src/test/java/org/finos/legend/engine/persistence/components/RelationalMultiDatasetIngestorTest.java [hunk 9]
```diff
@@ -672,12 +683,38 @@ public void testMixedIngestMode() throws IOException
...
+                .group(testSchemaName)
+                .name(dataset3 + suffixForFinalTable)
+                .build();
+
         // Configure ingest stages
         IngestStage ingestStage1ForDataset1 = IngestStage.builder().ingestMode(bulkLoad1).stagingDataset(bulkLoadStageTableForDataset1).mainDataset(bulkLoadMainTableForDataset1).build();
         IngestStage ingestStage1ForDataset2 = IngestStage.builder().ingestMode(bulkLoad2).stagingDataset(bulkLoadStageTableForDataset2).mainDataset(bulkLoadMainTableForDataset2).build();
         IngestStage ingestStage2ForDataset2 = IngestStage.builder().ingestMode(nontemporalSnapshot).stagingDataset(nontemporalSnapshotStageTableForDataset2).mainDataset(nontemporalSnapshotMainTableForDataset2).stagingDatasetBatchIdField(batchIdName).build();
+        IngestStage ingestStage1ForDataset3 = IngestStage.builder().ingestMode(noOp).stagingDataset(noOpStageTable1ForDataset3).mainDataset(noOpMainTable1ForDataset3).build();
+        IngestStage ingestStage2ForDataset3 = IngestStage.builder().ingestMode(noOp).stagingDataset(noOpStageTable1ForDataset3).mainDataset(noOpMainTable2ForDataset3).build();
```
Comment (kumuwu) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-h2/src/test/java/org/finos/legend/engine/persistence/components/RelationalMultiDatasetIngestorTest.java:713]: the staging dataset for stage 2 should be `noOpStageTable2ForDataset3`

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-test/src/test/java/org/finos/legend/engine/persistence/components/scenarios/NoOpScenarios.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2022 Goldman Sachs
```
Comment (kumuwu) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-test/src/test/java/org/finos/legend/engine/persistence/components/scenarios/NoOpScenarios.java:1]: typo

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-test/src/test/java/org/finos/legend/engine/persistence/components/testcases/ingestmode/NoOpTestCases.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
+// Copyright 2022 Goldman Sachs
```
Comment (kumuwu) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-test/src/test/java/org/finos/legend/engine/persistence/components/testcases/ingestmode/NoOpTestCases.java:1]: typo

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/ingestmode/NoOpAbstract.java (hunk 6 above)
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/ingestmode/NoOpAbstract.java]: Fixed

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/NoOpPlanner.java (hunk 7 above)
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/NoOpPlanner.java]: Fixed

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-ansi/src/test/java/org/finos/legend/engine/persistence/components/ingestmode/NoOpTest.java (hunk 8 above)
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-ansi/src/test/java/org/finos/legend/engine/persistence/components/ingestmode/NoOpTest.java]: Fixed

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-h2/src/test/java/org/finos/legend/engine/persistence/components/RelationalMultiDatasetIngestorTest.java (hunk 9 above)
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-h2/src/test/java/org/finos/legend/engine/persistence/components/RelationalMultiDatasetIngestorTest.java]: Fixed

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-test/src/test/java/org/finos/legend/engine/persistence/components/scenarios/NoOpScenarios.java (hunk 10 above)
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-test/src/test/java/org/finos/legend/engine/persistence/components/scenarios/NoOpScenarios.java]: Fixed

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-test/src/test/java/org/finos/legend/engine/persistence/components/testcases/ingestmode/NoOpTestCases.java (hunk 11 above)
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-test/src/test/java/org/finos/legend/engine/persistence/components/testcases/ingestmode/NoOpTestCases.java]: Fixed

**Review by prasar-ashutosh — APPROVED** [review]

**Review by gs-ssh16 — APPROVED** [review]

**Pull request merged**